stan/math/prim/functor/apply_scalar_ternary.hpp

```diff
@@ -0,0 +1,216 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_APPLY_SCALAR_TERNARY_HPP
+#define STAN_MATH_PRIM_FUNCTOR_APPLY_SCALAR_TERNARY_HPP
+
+#include <stan/math/prim/fun/as_column_vector_or_scalar.hpp>
+#include <stan/math/prim/meta/is_stan_scalar.hpp>
+#include <stan/math/prim/meta/is_container.hpp>
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <stan/math/prim/err/check_matching_dims.hpp>
+#include <stan/math/prim/err/check_matching_sizes.hpp>
+#include <stan/math/prim/fun/num_elements.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Base template function for vectorization of ternary scalar functions
+ * defined by applying a functor to a combination of scalars,
+ * containers of matching sizes, or a combination of a scalar and a container.
+ * These containers can be a standard library vector, Eigen dense
+ * matrix expression template, or container of these. For each specialisation,
+ * the same type as the largest (dimension) input is returned.
+ *
+ * This base template function takes (and returns) scalars.
+ *
+ * @tparam T1 Type of first scalar to which functor is applied.
+ * @tparam T2 Type of second scalar to which functor is applied.
+ * @tparam T3 Type of third scalar to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x First input to which operation is applied.
+ * @param y Second input to which operation is applied.
+ * @param z Third scalar to which operation is applied.
+ * @param f functor to apply to inputs.
+ * @return Scalar with result of applying functor to inputs.
+ */
+template <typename T1, typename T2, typename T3, typename F,
+          require_all_stan_scalar_t<T1, T2, T3>* = nullptr>
+inline auto apply_scalar_ternary(const T1& x, const T2& y, const T3& z,
+                                 const F& f) {
+  return f(x, y, z);
+}
```

Comment:
Why is `f` in the back?

Reply:
This was a hold-over from the `apply_scalar_binary` structure which had the functor as the last value. Which I now regret, and will probably change in another PR!

Changed for `apply_scalar_ternary` in this PR